### ISSUE TITLE
Different class layout for LayoutKind.Explicit in .NET 7

### ISSUE THREAD

**Issue by f2bo** [dotnet/runtime#80973]
### Description

I've a class library targetting .NET 6 that wraps a native API using P/Invoke. I also have a test project for this library that was also targetting .NET 6 until recently and where several tests began to fail after I switched its target framework to .NET 7. I traced the cause to a change in behavior in .NET 7 for a class with an explicit layout. The notable aspect of this class is that it has a base class. It seems that .NET 7 adds additional padding for the base class that it did not include in .NET 6 so that the derived class members are not at the expected locations.

Note that I've previously encountered unexpected behavior with fixed layouts when calling a native function with an instance of a (non-blittable) derived class whenever the data type of the parameter was declared as the (blittable) base class type. In that case, the parameter would be pinned instead of marshalled and the behavior was incorrect. However, this does not seem to apply in this case. 

### Reproduction Steps
UPDATE: The code below is a minimal example that reproduces the issue. It is NOT the actual native code which involves a union. See https://github.com/dotnet/runtime/issues/80973#issuecomment-1399302883 for additional information.

- Create a native library with the folllowing code:

```C
#define EXPORT extern "C" __declspec (dllexport)

typedef struct {
    int type;
    int version;
    int other;
} Config;

EXPORT void Configure(Config* config)
{
    config->type = 2;
    config->version = 99999;
}
```
- Create a  .NET console application with the following code:

```C#
// LayoutKind.Explicit + FieldOffset(0) for Type has the same behavior
[StructLayout(LayoutKind.Sequential)]
public abstract class ConfigCommon
{
    public int Type;
}

[StructLayout(LayoutKind.Explicit)]
public class Config : ConfigCommon
{
    [FieldOffset(0)] public int Version;
    [FieldOffset(4)] public int Other;
}

static void Main()
{
    Config config = new()
    {
        Type = 1,
        Version = 11111
    };

    Configure(config);
    Console.WriteLine($"{RuntimeInformation.FrameworkDescription} -  Type: {config.Type}, Version: {config.Version}");

    [DllImport(...)]   // set the name to the native library created previously
    static extern void Configure([In, Out] Config config);
}

```
- Set the target framework to .NET 6, clean, build and run. It displays:
```
.NET 6.0.13 -  Type: 2, Version: 99999
```
- Change the target framework to .NET 7, clean, build, and run. It displays:
```
.NET 7.0.2 -  Type: 2, Version: 11111
```
Note that the Version property was not set correctly.

### Expected behavior

No padding should be added to base class when computing the offset of the derived class members.

### Actual behavior

Inspecting the parameter received by the native function in the debugger shows the differences in layout between .NET 6 and .NET 7.

In .NET 6, the `version` member appears at offset 4, as expected.

![image](https://user-images.githubusercontent.com/1148767/213870102-67527d7a-aefc-4b9a-a9c6-a8677ec6d4b0.png)

In .NET 7, the `version` member appears to have been written at offset 8 where the `other` member was expected.

![image](https://user-images.githubusercontent.com/1148767/213870116-a046e603-8447-456c-9b26-9b2ef718cafc.png)

### Regression?

Apparently. This works in .NET 6.03.

### Configuration

.NET SDK:
 Version:   7.0.102
 Commit:    4bbdd14480

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.102\

Host:
  Version:      7.0.2
  Architecture: x64
  Commit:       d037e070eb

.NET SDKs installed:
  2.1.526 [C:\Program Files\dotnet\sdk]
  3.1.426 [C:\Program Files\dotnet\sdk]
  5.0.408 [C:\Program Files\dotnet\sdk]
  6.0.101 [C:\Program Files\dotnet\sdk]
  6.0.113 [C:\Program Files\dotnet\sdk]
  6.0.308 [C:\Program Files\dotnet\sdk]
  7.0.100 [C:\Program Files\dotnet\sdk]
  7.0.102 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

### Other information

Just in case, I've also tried setting several combinations of the `StructLayout(..., Pack=...)` but it does not appear to change the outcome.

**Comment by tannergooding** [comment]
Outside the issue you've presented, it is worth calling out that a native layout of:
```cpp
typedef struct {
    int type;
    int version;
    int other;
} Config;
```

and a managed layout of:
```csharp
[StructLayout(LayoutKind.Sequential)]
public abstract class ConfigCommon
{
    public int Type;
}

[StructLayout(LayoutKind.Explicit)]
public class Config : ConfigCommon
{
    [FieldOffset(0)] public int Version;
    [FieldOffset(4)] public int Other;
}
```

Are not compatible. The native definition is not using a `union` or any kind of field overlapping. It likewise is not using any form of inheritance (which can impact field layout), etc.

The following managed definitions would be compatible:
```csharp
public struct Config
{
    public int type;
    public int version;
    public int other
}
```

When the native side takes a pointer (`Config*`) it is recommended you either:
* Match this on the managed side, passing along `Config*`
* Use the relevant `ref Config`, `out Config`, or `in Config` + the `LibraryImport` source generator

Noting for the latter that which of `ref`/`out`/`in` is correct depends on how the C/C++ function handles the data. Also noting that you can also use regular P/Invoke, but that depends on the built-in runtime marshalling and new code ideally moves away from this in favor of `LibraryImport` or directly writing blittable code.

You can also technically use `[StructLayout(LayoutKind.Sequential)] public class Config { public int type; public int version; public int other; }` for some `Config*`. However, there are many downsides to this approach and I wouldn't recommend it, especially for new code.

**Comment by tannergooding** [comment]
For the issue presented, the "expected behavior" presented is not correct:
> No padding should be added to base class when computing the offset of the derived class members.

Consider a native type defined as:
```cpp
struct Base {
    int64_t a;
    int32_t b;
};

struct Derived : Base {
    int32_t c;
};
```

On a 64-bit system, the size and alignment of `Base` is `16` and `8`, respectively (because of the `int64_t` field). `Derived` then inherits this alignment and adds its own fields starting at `offset 16`. This is different from if `c` were directly defined as part of `Base`, in which case it would be at `offset 12`

In this particularly case, the managed alignment of your `Config` setup is `8` and I expect that the marshaller is not adjusting for the fact that the "hidden fields" that are part of managed object allocations should not impact the "native layout" when the type is passed to unmanaged.

If it had adjusted the alignment, the computed size/layout would be `8`. However, it is illegal in C for a `union` to be the base type of another or to itself have a base type. That is, the closest in native to what you've defined in managed is:
```cpp
struct ConfigCommon {
    int type;
};

union Config : ConfigCommon {
    int version;
    int other;
};
```

However, such a type is illegal in C and results in a compiler error. So it is not particularly relevant that this edge case be handled.

**Comment by f2bo** [comment]
Hi @tannergooding. Thank you for your reply. 

I intentionally excluded some information from the reproduction steps above that only added noise and did not change the problem I wanted to describe. The native definition is in fact a union with fields that overlap. 

The native API has a number of functions that take a `Config*` parameter and, as you pointed out, does not require inheritance, in which case I could have just used a single struct rather than a class. Each of the union elements below has a `version` as its first member.

```C
struct Config {
    PayloadType payload_type;

    union {
        VideoConfig video_config;
        AudioConfig audio_config;
        AncillaryDataConfig ancillary_data_config;
    };
};
```
However, a more recent version of the native API allows for extensibility and therefore adds a number of functions that instead take a `ConfigCommon*` parameter with only a `payload_type` and `version` members, allowing new custom types to be introduced. There is a mechanism that allows registering these custom configurations by providing a vTable with functions that handle them.  This is the reason I originally modelled the managed API as a class with inheritance instead of a struct, and it was working correctly under .NET 6, at least in the way that I expected. 

```C
typedef struct {
    PayloadType payload_type;
    ProfileVersion version;
} ConfigCommon;
```

Now, going back to the actual problem. 

>On a 64-bit system, the size and alignment of Base is 16 and 8, respectively (because of the int64_t field). Derived then inherits this alignment and adds its own field (c) starting at offset 16. This is different from if c were directly defined as part of Base and no inheritance were involve, in which case c would be at offset 12

This may be the problem as I assumed that by specifying an explicit layout and`[FieldOffset(0)]`, it would cause the derived class members to be placed immediately following the base class members, as if they were part of the same class. Or rather, when I originally looked at this, I couldn't find anything in the documentation indicating whether this was accurate but instead based my assumption on the actual behavior that I observed under .NET 6. Perhaps the problem is in .NET 6 and it worked to my advantage. I don't know.

> In this particularly case, the managed alignment of your Config setup is 8 and I expect that the marshaller is not adjusting for the fact that the "hidden fields" that are part of managed object allocations should not impact the "native layout" when the type is passed to unmanaged.

I'm confused by this statement but I'm probably misunderstanding what you mean. I thought that the managed layout was the same as the native layout in this particular case, as everything is blittable, and that the object was simply being pinned before being passed to the native code. In fact, I can successfully allocate a pinned GCHandle for a `Config` instance. Where and how is the marshaller involved?

Just to be clear, I'm not blocked by this issue as I have alternatives such as changing ConfigCommon to a memberless class with abstract virtual properties or an interface and have the derived classes implement them. This will require some other changes but it's doable. I just wanted to report the change in behavior.

**Comment by tannergooding** [comment]
> Each of the union elements below has a version as its first member.

The correct managed definition for this that will work across all platforms is:
```csharp
public struct Config
{
    public PayloadType payload_type;

    public _Anonymous_e__Union Anonymous;

    [StructLayout(LayoutKind.Explicit)]
    internal partial struct _Anonymous_e__Union
    {
        [FieldOffset(0)]
        public VideoConfig video_config;

        [FieldOffset(0)]
        public AudioConfig audio_config;

        [FieldOffset(0)]
        public AncillaryDataConfig ancillary_data_config;
    }

    // These next three members are "optional" and just make it friendlier to use

    [UnscopedRef]
    public ref VideoConfig video_config => ref Anonymous.video_config;

    [UnscopedRef]
    public ref AudioConfig audio_config => ref Anonymous.audio_config;

    [UnscopedRef]
    public ref AncillaryDataConfig ancillary_data_config => ref Anonymous.ancillary_data_config;
}
```

At the C/C++ level, "anonymous structs/unions" are not actually "inline" but rather distinct and separate `record declarations`. This *can* impact padding on some platforms or for specific types of field combinations and so trying to declare the members inline in .NET may not be equivalent. It may likewise not be possible if one of the preceding fields is a pointer or other variable sized type, so it is always best to do the correct thing.

Something appearing to work on one platform doesn't strictly mean that it will work on or be compatible with all platforms that the code could run against.

> I'm confused by this statement but I'm probably misunderstanding what you mean. I thought that the managed layout was the same as the native layout in this particular case, as everything is blittable, and that the object was simply being pinned before being passed to the native code. In fact, I can successfully allocate a pinned GCHandle for a Config instance. Where and how is the marshaller involved?

Classes are not blittable, even if they only contain blittable types. They likewise are not purely the fields that you define like a struct would be. In particular, every managed reference type has some hidden fields such as the `Method Table Pointer`. This field is not included in the part of the data that is marshalled to native, but such fields can impact the managed layout and therefore can impact packing/alignment of data contained within. Likewise just having a reference type in your class/struct can likewise impact the layout as there are strict requirements around how managed references exist in memory.

In general it is recommended that you try to match the native definition of the type as closely as possible. Doing otherwise can lead to subtle bugs where the types do not match. This is often a side effect of mismatched layout making incorrect assumptions that do not apply to all target platforms.

**Comment by AaronRobinsonMSFT** [comment]
> I just wanted to report the change in behavior.

@f2bo Thank you for that.

Also, thanks to @tannergooding for the excellent explanation of the correct way to do this.

> In general it is recommended that you try to match the native definition of the type as closely as possible. Doing otherwise can lead to subtle bugs where the types do not match. This is often a side effect of mismatched layout making incorrect assumptions that do not apply to all target platforms.

+1

@jkoritzinsky Thoughts on the above issue? I don't see any action on our side but open to options.

**Comment by jkoritzinsky** [comment]
Given that the idea of representing unions as classes has come up multiple times, we should probably provide guidance in the docs about/against it and possibly provide a sample of how to correctly do it with LibraryImportGenerator. That might help alleviate some of the pain here in the future.

**Comment by f2bo** [comment]
@tannergooding Thank you for the explanation and for providing the managed type definition. 

Currently, the managed library exposes a public `Config` class as well as an abstract `ConfigCommon` base class that users need to derive from in order to use custom configurations in v2 of the native API. The definition you propose eliminates this hierarchy, so I assume the intention is for this struct to only be used internally and for there to exist a custom stage that marshals these public classes into your struct definition. In other words, the approach taken by the LibraryImportGenerator. I did consider using the generator while it was still under development but had some issues (https://github.com/dotnet/runtime/issues/69096) and decided to wait until the design evolved.  

Before I saw your last message, I had already implemented a quick fix by removing all fields from the `ConfigCommon` base class, but still keeping the class hierarchy, as shown below. 

```C#
[StructLayout(LayoutKind.Explicit)]
public abstract class ConfigCommon
{
    public abstract PayloadType payload_type { get; set; }
    public abstract ProfileVersion version { get; set; }
}

[StructLayout(LayoutKind.Explicit)]
public class Config : ConfigCommon
{
    [FieldOffset(0)] private PayloadType _payload_type;
    [FieldOffset(4)] public VideoConfig video_config;
    [FieldOffset(4)] public AudioConfig audio_config;
    [FieldOffset(4)] public AncillaryDataConfig ancillary_data_config;

    public override CdiBaselineAvmPayloadType payload_type { get => _payload_type; set => _payload_type = value; }
    public override CdiAvmBaselineProfileVersion version { get => video_config.version; set => video_config.version = value; }
}
```

This required the least amount of change and seems to work with both .NET 6 and .NET 7 clients. However, I do get the message that the preferred approach is to use structs and custom marshalling.  The library still targets .NET 6, but the plan is to eventually move to .NET 7 and begin using the LibraryImportGenerator, so making the changes necessary to take advantage of your definition seems worthwhile and a step in that direction. 

> Given that the idea of representing unions as classes has come up multiple times, we should probably provide guidance in the docs about/against it ...

@jkoritzinsky The problem, at least in my case, is that the documentation for `StructLayout` does not indicate the behavior when applied to a class hierarchy, either to forbid it or to warn that an explicit layout could turn out to be not so explicit depending on the version of the runtime.
